aesd-char-driver/main.c

```diff
@@ -191,6 +255,67 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff
     return retval;
 }
 
+////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////
+static loff_t aesd_seek(struct file* filp, loff_t f_pos, int whence){
+	
+	PDEBUG("-seek to offset %lli from %d",f_pos, whence);
+	
+	//Get our dev structure
+	struct aesd_dev *dev = (struct aesd_dev *)filp->private_data;
+	
+	loff_t newPos = 0;
+	size_t dataSize = 0;
+	
+	//First we need to get the size of the data set, which we'll use for SEEK_END, and 
+	//	range checking at the end.
+	PDEBUG("Getting the mutex lock");
+	if(mutex_lock_interruptible(&dev->mutex))
+		return -ERESTARTSYS;
+	
+	PDEBUG("Calculating current length");
+	uint8_t index;
+	struct aesd_buffer_entry *entry;
+	AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->cirBuf, index)
+		dataSize += entry->size;	
+	
+	
+	//Now we'll set the newPos value based on where the user wants us to move to.
+	switch(whence){
+		case SEEK_SET:
+			newPos = f_pos;
+			break;
+		case SEEK_CUR:
+			newPos = filp->f_pos + f_pos;
+			break;
+		case SEEK_END:
+			newPos = dataSize + f_pos;
+			break;		
+		default:
+			//Here we'll just set to -1 so we'll return EINVAL after we release the mutex
+			newPos = -1;
+	}
+	
+	PDEBUG("User wishes to seek to %lli", newPos);
+	
+	//Check if the newPos is valid.
+	if(newPos < 0 || newPos > dataSize){
+		PDEBUG("That position is invalid: < 0 or > %li", dataSize);
+		newPos = -EINVAL;
+	}
+	else{
+		filp->f_pos = newPos;
+	}
+
+	PDEBUG("Unlocking the mutex");
+	mutex_unlock(&dev->mutex);
```

Comment:
@JLDenn see discussion in class, this mutex unlock can be moved before the switch statement.